services/streammanager/kafka/kafkamanager.go

```diff
@@ -354,78 +369,72 @@ func Produce(jsonData json.RawMessage, producer interface{}, destConfig interfac
 	err := json.Unmarshal(jsonData, &batch)
 	if err != nil {
-		return 400, "Failure", "Error while unmarshalling json data :: " + err.Error()
+		return 400, "Failure", "Error while unmarshalling json data: " + err.Error()
 	}
 
-	batchedMessage, err := prepareBatchedMessage(topic, batch, timestamp)
+	timestamp := time.Now()
+	batchOfMessages, err := prepareBatchOfMessages(topic, batch, timestamp)
 	if err != nil {
-		return 400, "Failure", "Error while preparing batched message :: " + err.Error()
+		return 400, "Failure", "Error while preparing batched message: " + err.Error()
 	}
 
-	err = kafkaProducer.SendMessages(batchedMessage)
+	err = p.Publish(context.TODO(), batchOfMessages...)
 	if err != nil {
 		return makeErrorResponse(err) // would retry the messages in batch in case brokers are down
 	}
 
 	returnMessage := "Kafka: Message delivered in batch"
-	statusCode := 200
-	errorMessage := returnMessage
-
-	return statusCode, returnMessage, errorMessage
+	return 200, returnMessage, returnMessage
 }
 
-func sendMessage(jsonData json.RawMessage, kafkaProducer sarama.SyncProducer, topic string) (int, string, string) {
-	timestamp := time.Now()
+func sendMessage(jsonData json.RawMessage, p producer, topic string) (int, string, string) {
 	parsedJSON := gjson.ParseBytes(jsonData)
-	data := parsedJSON.Get("message").Value().(interface{})
+	messageValue := parsedJSON.Get("message").Value()
+	if messageValue == nil {
+		return 400, "Failure", "Invalid message"
+	}
+
+	data := messageValue.(interface{})
 	value, err := json.Marshal(data)
 	if err != nil {
 		return makeErrorResponse(err)
 	}
-	userID, _ := parsedJSON.Get("userId").Value().(string)
 
+	timestamp := time.Now()
+	userID, _ := parsedJSON.Get("userId").Value().(string)
 	message := prepareMessage(topic, userID, value, timestamp)
-
-	partition, offset, err := kafkaProducer.SendMessage(message)
-	if err != nil {
+	if err = p.Publish(context.TODO(), message); err != nil {
 		return makeErrorResponse(err)
 	}
 
-	returnMessage := fmt.Sprintf("Message delivered at Offset: %v , Partition: %v for topic: %s", offset, partition, topic)
-	statusCode := 200
-	errorMessage := returnMessage
-
-	return statusCode, returnMessage, errorMessage
+	returnMessage := fmt.Sprintf("Message delivered to topic: %s", topic)
+	return 200, returnMessage, returnMessage
 }
 
 func makeErrorResponse(err error) (int, string, string) {
-	returnMessage := fmt.Sprintf("%s error occured.", err.Error())
-	statusCode := getStatusCodeFromError(err) //400
-	errorMessage := err.Error()
+	returnMessage := fmt.Sprintf("%s error occurred.", err)
 	pkgLogger.Error(returnMessage)
-	return statusCode, returnMessage, errorMessage
+	return getStatusCodeFromError(err), returnMessage, err.Error()
 }
 
 // getStatusCodeFromError parses the error and returns the status so that event gets retried or failed.
 func getStatusCodeFromError(err error) int {
 	statusCode := 500
-
 	errorString := err.Error()
 
 	for _, s := range abortableErrors {
```

Comment:
shouldn't we adapt `abortableErrors` as well?
They should be different, e.g. here is what sarama has:
https://github.com/Shopify/sarama/blob/f07b7b87556741de750337199ccd02aa2a996fe2/errors.go
And here are segment's:
https://github.com/segmentio/kafka-go/blob/main/error.go

Shouldn't we abstract this as well on our generic client?

Reply:
I'm thinking of removing our own `abortableErrors` slice and do this instead:

```go
// ...
if client.IsProducerErrTemporary(err) {
    statusCode = 400
    break
}
// ...
```

Then add the `isTemporary()` method to the producer and use the same logic as the segment lib retrier:

```go
var tempError interface{ Temporary() bool }

func IsProducerErrTemporary(err error) bool {
	isTransientNetworkError := errors.Is(err, io.ErrUnexpectedEOF) ||
		errors.Is(err, syscall.ECONNREFUSED) ||
		errors.Is(err, syscall.ECONNRESET) ||
		errors.Is(err, syscall.EPIPE)
	if isTransientNetworkError {
		return true
	} 
	if errors.As(err, &tempError) {
		return tempError.Temporary()
	}
	return false
}
```

I think it would make sense to remove `MaxRetries` as well and use only one retry logic. Thoughts?